File: laboratorios/Lab2/Parte1/Fibonacci/Fibonacci.java
package algoritmos_lab1;

public class Fibonacci {

	public static double phi1 = (1.0 + Math.sqrt(5.0))/2.0;
	public static double phi2 = (1.0 - Math.sqrt(5.0))/2.0;

	public static long fiboLinear(int n) {
		int i;
		long fibo0, fibo1, fibo;
		if(n == 0) 			
			return 0;
		else if(n == 1) 	
			return 1;
		fibo = 0;
		fibo0 = 0; 
		fibo1 = 1; 
		for(i=1; i<n; i++) {
			fibo = fibo0 + fibo1; 
			fibo0 = fibo1; 
			fibo1 = fibo;
		}
		return fibo;
	}

	public static long fiboRecursive(int n) {
		if(n<2) return (long) n;
		else  	return fiboRecursive(n-1) + fiboRecursive(n-2);
			
	}

	public static long fiboAnalythic(int n) {
		return (long) ((long) (Math.pow(phi1, n) - Math.pow(phi2, n))/Math.sqrt(5.0));
	}

	public static void testByte() {
		int n=0;
		byte fibo_test=0;
		long t0, t1, fibo_real=0;
		// Test lineal:
		t0 = System.nanoTime();
		while(fibo_real >= 0 && fibo_real < Math.pow(2, 7)) {
			fibo_real = fiboLinear(n);
			fibo_test = (byte) fiboLinear(n);
			//System.out.println("\t- n: " + n + ", real: " + fibo_real + ", test: " + fibo_test);
			n++;
		}
		t1 = System.nanoTime();
		System.out.println("- Overflow Bytes Fibonacci: (N = " + (n-1) + ", t=" + (t1 - t0) + " ns, avg=" + (t1 - t0)/((double) n) + " ns)");
	}

	public static void testShort() {
		int n=0;
		short fibo_test=0;
		long t0, t1, fibo_real=0;
		// Test lineal:
		t0 = System.nanoTime();
		while(fibo_real >= 0 && fibo_real < Math.pow(2, 15)) {
			fibo_real = fiboLinear(n);
			fibo_test = (short) fiboLinear(n);
			//System.out.println("\t- n: " + n + ", real: " + fibo_real + ", test: " + fibo_test);
			n++;
		}
		t1 = System.nanoTime();
		System.out.println("- Overflow Short Integer Fibonacci: (N = " + (n-1) + ", t=" + (t1 - t0) + " ns, avg=" + (t1 - t0)/((double) n) + " ns)");
	}

	public static void testInt() {
		int n=0;
		int fibo_test=0;
		long t0, t1, fibo_real=0;
		// Test lineal:
		t0 = System.nanoTime();
		while(fibo_real >= 0 && fibo_real < Math.pow(2, 31)) {
			fibo_real = fiboLinear(n);
			fibo_test = (int) fiboLinear(n);
			//System.out.println("\t- n: " + n + ", real: " + fibo_real + ", test: " + fibo_test);
			n++;
		}
		t1 = System.nanoTime();
		System.out.println("- Overflow Integer Fibonacci: (N = " + (n-1) + ", t=" + (t1 - t0) + " ns, avg=" + (t1 - t0)/((double) n) + " ns)");
	}

	public static void testLong() {
		int n=0;
		long t0, t1, fibo_real=0, fibo_test=0;
		// Test lineal:
		t0 = System.nanoTime();
		while(fibo_real >= 0 && fibo_real < Math.pow(2, 63)) {
			fibo_real = fiboLinear(n);
			fibo_test = fiboLinear(n);
			//System.out.println("\t- n: " + n + ", real: " + fibo_real + ", test: " + fibo_test);
			n++;
		}
		t1 = System.nanoTime();
		System.out.println("- Overflow Long Fibonacci: (N = " + (n-1) + ", t=" + (t1 - t0) + " ns, avg=" + (t1 - t0)/((double) n) + " ns)");
	}
	
	public static void main(String[] args) {
		// Test Byte
		testByte();
		// Test Short
		testShort();
		// Test int
		testInt();
		// Test long
		testLong();
	}

}

File: laboratorios/Lab2/Parte1/Fibonacci/Fibonacci.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

const double PHI1 = (1.0 + sqrt(5.0))/2.0;
const double PHI2 = (1.0 - sqrt(5.0))/2.0;

unsigned long long fiboLinear(int n) {
    int i;
    unsigned long long fibo0, fibo1, fibo;
    if(n == 0) 			
        return 0;
    else if(n == 1) 	
        return 1;
    fibo = 0;
    fibo0 = 0; 
    fibo1 = 1; 
    for(i=1; i<n; i++) {
        fibo = fibo0 + fibo1; 
        fibo0 = fibo1; 
        fibo1 = fibo;
    }
    return fibo;
}

unsigned long long fiboRecursive(int n) {
    if(n<2) return (unsigned long long) n;
    else  	return fiboRecursive(n-1) + fiboRecursive(n-2);
        
}

unsigned long long fiboAnalythic(int n) {
    return (unsigned long long) ((unsigned long long) (pow(PHI1, n) - pow(PHI2, n))/sqrt(5.0));
}

void testChar() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(char)-1))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Char Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testUnsignedChar() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(char)))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Unsigned Char Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/(((double) n)*CLOCKS_PER_SEC));
}

void testShort() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(short)-1))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Short Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testUnsignedShort() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(short)))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Unsigned Short Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testInt() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(int)-1))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Integer Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testUnsignedInt() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(int)))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Unsigned Integer Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testLong() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(long)-1))
        fibo = fiboLinear(n++);
    t1 = clock();
    printf("- Overflow Long Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

void testUnsignedLong() {
    int n=0;
    time_t t0, t1;
    unsigned long long fibo=0, fibo0=0;
    // Test lineal:
    t0 = clock();
    while(fibo < pow(2, 8*sizeof(long)) && fibo >= fibo0) {
        fibo0 = fibo;
        fibo = fiboLinear(n++);
    }
    t1 = clock();
    printf("- Overflow Unsigned Long Fibonacci: (N=%d, t=%f ns, avg=%lf ns)\n", n-1, (float) (t1 - t0)/CLOCKS_PER_SEC, (t1 - t0)/((double) n));
}

int main() {
    // Test Char
    testChar();
    // Test Unsigned Char
    testUnsignedChar();
    // Test Short
    testShort();
    // Test Unsigned Short
    testUnsignedShort();
    // Test int
    testInt();
    // Test unsigned int
    testUnsignedInt();
    // Test long
    testLong();
    // Test unsigned long
    testUnsignedLong();
}
